### ISSUE TITLE
Token type collision

### ISSUE THREAD

**Issue by toniperic** [friendsoftwig/twigcs#107]
I've tried printing the token stream for the following template:
```twig
{% if hello and
    world %}
{% endif %}
```

This was the result:
```
PHP Fatal error:  Method Twig\Token::__toString() must not throw an exception, caught LogicException: Token of type "13" does not exist.
```

Additionally, if I try to do the same for the following template:
```twig
{% if hello %}
{% endif %}
```

This is the result:
```
BLOCK_START_TYPE()
ARROW_TYPE( )
NAME_TYPE(if)
ARROW_TYPE( )
NAME_TYPE(hello)
ARROW_TYPE( )
BLOCK_END_TYPE()
BLOCK_START_TYPE()
ARROW_TYPE( )
NAME_TYPE(endif)
ARROW_TYPE( )
BLOCK_END_TYPE()
```

The `ARROW_TYPE` seemed odd to me, so after digging a bit it looks to me as the issue lies in the constants used by TwigCS which have conflicts with Twig.

More precisely, [here](https://github.com/friendsoftwig/twigcs/blob/7c4bc76dad8d851045b9eeab4979a240550f8300/src/Token.php#L7) is defined a constant which should represent whitespace, but eventually [Twig caught up](https://github.com/twigphp/Twig/blob/1bce2f30511c1aff73572ad2338d0a612f207621/src/Token.php#L37) with the numbering.

So these two seem to be kinda connected.

As a reference, I was toying around with `v3.2.2`.

**Comment by toniperic** [comment]
Just gonna add some more info related to this.

Template as a reference point:
```twig
{% set foobar = foo|map(bar => baz) %}
```

Tried using the Official ruleset on this, the following was the result:
> l.1 c.29 : ERROR There should be 1 space(s) before ">".

When I printed out tokens, the following was the result:
```
BLOCK_START_TYPE()
ARROW_TYPE( )
NAME_TYPE(set)
ARROW_TYPE( )
NAME_TYPE(foobar)
ARROW_TYPE( )
OPERATOR_TYPE(=)
ARROW_TYPE( )
NAME_TYPE(foo)
PUNCTUATION_TYPE(|)
NAME_TYPE(map)
PUNCTUATION_TYPE(()
NAME_TYPE(bar)
ARROW_TYPE( )
OPERATOR_TYPE(=)
OPERATOR_TYPE(>)
ARROW_TYPE( )
NAME_TYPE(baz)
PUNCTUATION_TYPE())
ARROW_TYPE( )
BLOCK_END_TYPE()
```

**Comment by toniperic** [comment]
I've also found out about #69 but I'm not sure whether that means versions older than 4 would not be getting the arrow token support?

**Comment by OwlyCode** [comment]
The easiest way to fix this would be to change the numbers of the custom tokens Twigcs introduces but it feels like it's just a workaround. This problem lies inside a greater problem I'm trying to figure out how to solve: keeping up with twig versions in a consistent way.

Nowadays I'm tempted to go with some opinionated choices:

- Stripping the Lexer from the official library, we copy it inside Twigcs and maintain it separately with our needed code (custom tokens, newlines handling): no more collisions, no more twig dependency inside composer, we gain full control on that.
- Killing twigcs 3.x as fast as possible (so many unfixable things like arrow functions)
- Adding Twig 1, 2 and 3 support inside twigcs 4.x before it is out of beta (using 3 rulesets ?)

**Comment by OwlyCode** [comment]
I've submitted a PR that will fix this issue here : #114.

**Comment by OwlyCode** [comment]
I'm closing this as the PR got merged!
